add migration doc from wiki contents (#747)
canonical/ubuntu-server-documentation: docs/explanation/virtualisation/live-migration.md

- QEMU acts as the engine, actively streaming the VM's live memory (RAM) and CPU state across the network. It can work without libvirt. A live migration can be triggered using QEMU Monitor Protocol (QMP).

- libvirt acts as the manager, coordinating the setup, security, and handoff between the two physical servers so the guest operating system and its applications keep running without interruption.

Although migration is a complex process, it generally works well if you migrate VMs between source and destination hosts with very similar capabilities, both on hardware (CPU model) and software (QEMU version) levels. But the reality is that the infrastructure is very often heterogeneous, with CPUs and virtualisation software stacks at different generations or versions. Migration can still happen across hosts with different hardware and software versions (for example, different QEMU versions). Two mechanisms make this possible:

- **Versioned machine types**, which keep the guest's virtual hardware layout stable across QEMU versions.
- **CPU model + named features**, which allows the CPU to be expressed as a baseline model name +/- named features.

## Versioned machine types

> The QEMU machine type defines the exact virtual hardware blueprint (chipset, PCI slots, ACPI tables) of a virtual machine. This allows QEMU to offer to virtual machines a well-defined and stabilized foundation they can run on so that they are not impacted by changes that happen under the hood on the virtualisation software stack. The machine types happen to be very useful to ensure live migration succeeds, especially inside heterogeneous infrastructure.

Reviewer: ```suggestion
```

Drop this as it is better explained in the extra page now.

> When you live-migrate a VM, you are taking a snapshot of a running system's RAM and CPU state and sending it over the wire. The destination host must have a virtual hardware environment that matches that state down to the very last byte; this is possible by keeping the machine type unchanged. However, when the QEMU software is upgraded, machine types can evolve and get new features that might be incompatible with the migrated virtual machine. The solution is to version the machine types by creating a new version of the machine type for each supported QEMU version.

Reviewer: ```suggestion
When you live-migrate a VM, you are taking a snapshot of a running system's RAM and CPU state and sending it over the wire. The destination host must have a virtual hardware environment that matches that state exactly; this is possible due to both sides of the migration agreeing on the same QEMU machine type.
```

Reviewer 2: What does "over the wire" mean? This feels quite colloquial

For more detail on Ubuntu machine types and how to upgrade the machine type of an existing guest, see {ref}`qemu-machine-types`.

## CPU baseline and named features

Versioned machine types keep the virtual platform stable, but they do not control the CPU that the guest sees. The CPU is exposed separately, and it is the second mechanism that makes migration across heterogeneous hosts possible.

> Libvirt and QEMU allow the CPU to be expressed in ways that ease migration between hosts in heterogeneous infrastructures. While a libvirt VM configured with `host-passthrough` or `max` (the equivalent QEMU `-cpu host` and `-cpu max`) takes the most out of the hardware and kernel hypervisor, it exposes the guests to all available host capabilities. Therefore, migration to another host is much less transparent to the guests (by exposing new capabilities or deprecating existing features), and this can break the guests after migration. Instead of specifying the CPU capabilities in such manner, a CPU can be expressed as a baseline CPU model +/- a list of features. In QEMU, it comes down to this example `-cpu` argument:

Reviewer: ```suggestion
Libvirt and QEMU allow the CPU to be expressed in ways that ease migration between hosts in heterogeneous infrastructures. While a libvirt VM configured with `host-passthrough` or `maximum` (the equivalent QEMU `-cpu host` and `-cpu max`) gets most out of the hardware, it exposes the guests to all available host capabilities. Therefore, migration to another host is much more restricted, because it needs to be able to present the very same environment to the guest. To do so it needs a clear definition that helps to ensure both sides of the migration can present the same. Therefore it is preferred to specifying the CPU capabilities as a baseline CPU model +/- a list of features to ensure compatibility. In QEMU, it comes down to this example `-cpu` argument:
```

Author: I struggle to be convinced by the sentence: "Instead of `max` to ensure compatibility it is therefore preferred to specifying the CPU capabilities ...", can you confirm that this is the text you like to propose, if yes, please help me there

Reviewer: Yes, that is what I'd suggest even when re-reading.
Maybe accept it as is and we wait for Sally's review next which will tear both our sentences apart for the greater good anyway?

Author: It is more about the technical meaning,
what do you mean by `max` to ensure compatibility ? which compatibility do you have in mind ?

```text
-cpu EPYC-Turin,x2apic=on,tsc-deadline=on,hypervisor=on,tsc-adjust=on,
 spec-ctrl=on,stibp=on,flush-l1d=on,ssbd=on,cmp-legacy=on,virt-ssbd=on,
 tsa-sq-no=on,tsa-l1-no=on,pcid=off,la57=off
```

> You can ask libvirt to build the CPU for you by using the `host-model` CPU mode:

Reviewer: ```suggestion
Or in a level higher, you can ask libvirt to build the CPU for you by using the `host-model` CPU mode.
This will detect the current CPU with all its features, but automatically expresses it as baseline CPU model +/- a list of features.
```

```xml
...
  <cpu mode='host-model' check='partial'/>
...
```

> or provide the exact CPU specification:

Reviewer: ```suggestion
or you might prefer even more control and provide the exact CPU specification:
```

```xml
...
  <cpu mode='custom' match='exact' check='full'>
    <model fallback='forbid'>EPYC-Turin</model>
    <vendor>AMD</vendor>
    <feature policy='require' name='x2apic'/>
    <feature policy='require' name='tsc-deadline'/>
    ...
    <feature policy='require' name='tsa-l1-no'/>
    <feature policy='disable' name='pcid'/>
    <feature policy='disable' name='la57'/>
...
```

> While the first approach seems easier and more appealing, it is not suitable for heterogeneous infrastructure. To retain full control over migratability across every host in the migration pool, the second approach is better. Explicitly specifying the CPU definition advertises to the guest only the CPU features that are guaranteed to exist on every host in the pool, rather than everything the current host happens to offer. This creates a stable "lowest common denominator" CPU that looks identical on both the source and the destination, so the guest never comes to rely on a feature that could disappear after a migration.

Reviewer: ```suggestion
While the first approach of just using `host-passthrough` and even `host-model` seems easier and more appealing, it is not suitable for heterogeneous infrastructure. To retain full control over migratability across every host in the migration pool a CPU should be modeled.
You can check the capabilities of every host in your pool and define the CPU definition to only advertise CPU features to the guest that are guaranteed to exist on every host in the pool.
This creates a stable "lowest common denominator" CPU that looks identical on both the source and the destination, so the guest never comes to rely on a feature that could disappear after a migration.
```

Author: When I mention "first approach", it is about using the "host-model" , not "host-passthrough", host-passthrugh has been definitely discarded earlier for migration consideration

Reviewer: That was not clear, but I'd not put host-model too far away.
Could is be ?

```
While the first approach of just using `host-passthrough` and even `host-model` seem easier and more appealing, it is not suitable for heterogeneous infrastructure. To retain full control over migratability across every host in the migration pool a CPU should be modeled to a common denominator.
```

Author: yes, I modified your suggestion and applied it. Thanks

### Determining the lowest common CPU denominator

> Finding the lowest denominator seems to be a complex task. Fortunately, libvirt provides a command to compute it (see {manpage}`virsh(1)` and its `hypervisor-cpu-baseline` command) by taking the CPU definitions of every host in the pool and returning a single baseline CPU that all of them can support.

Reviewer: Thank you for adding this section

- QEMU acts as the engine, actively streaming the VM's live memory (RAM) and CPU state across the network. It can work without libvirt. A live migration can be triggered using QEMU Monitor Protocol (QMP).

- libvirt acts as the manager, coordinating the setup, security, and handoff between the two physical servers so the guest operating system and its applications keep running without interruption.

Although migration is a complex process, it generally works well if you migrate VMs between source and destination hosts with very similar capabilities, both on hardware (CPU model) and software (QEMU version) levels. But the reality is that the infrastructure is very often heterogeneous, with CPUs and virtualization software stacks at different generations or versions. Migration can still happen across hosts with different hardware and software versions (for example, different QEMU versions). Two mechanisms make this possible:

- **Versioned machine types**, which keep the guest's virtual hardware layout stable across QEMU versions.
- **CPU model + named features**, which allows the CPU to be expressed as a baseline model name +/- named features.

## Versioned machine types

> The QEMU machine type defines the exact virtual hardware blueprint (chipset, PCI slots, ACPI tables) of a virtual machine. This allows QEMU to offer to virtual machines a well-defined and stabilized foundation they can run on so that they are not impacted by changes that happen under the hood on the virtualization software stack. The machine types happen to be very useful to ensure live migration succeeds, especially inside heterogeneous infrastructure.

Reviewer 2: ```suggestion
The QEMU machine type defines the exact virtual hardware blueprint (chipset, PCI slots, ACPI tables) of a virtual machine. This allows QEMU to offer VMs a well-defined and stabilized foundation they can run on, so they are not impacted by changes that happen "under the hood" on the virtualization software stack. The machine types are useful to ensure live migration succeeds, especially inside heterogeneous infrastructure.
```

When you live-migrate a VM, you are taking a snapshot of a running system's RAM and CPU state and sending it over the wire. The destination host must have a virtual hardware environment that matches that state down to the very last byte; this is possible by keeping the machine type unchanged. However, when the QEMU software is upgraded, machine types can evolve and get new features that might be incompatible with the migrated virtual machine. The solution is to version the machine types by creating a new version of the machine type for each supported QEMU version.

For more detail on Ubuntu machine types and how to upgrade the machine type of an existing guest, see {ref}`qemu-machine-types`.

## CPU baseline and named features

> Versioned machine types keep the virtual platform stable, but they do not control the CPU that the guest sees. The CPU is exposed separately, and it is the second mechanism that makes migration across heterogeneous hosts possible.

Reviewer 2: Unclear what you mean by "the second mechanism" (and by contrast, which one is the first)

Author: maybe "this is the second ..." is better ?

Reviewer 2: I'd prefer if we could be specific rather than vague, and actually state what mechanism we're referring to

> Libvirt and QEMU allow the CPU to be expressed in ways that ease migration between hosts in heterogeneous infrastructures. While a libvirt VM configured with `host-passthrough` or `maximum` (the equivalent QEMU `-cpu host` and `-cpu max`) gets most out of the hardware, it exposes the guests to all available host capabilities. Therefore, migration to another host is much more restricted, because it needs to be able to present the very same environment to the guest. To do so it needs a clear definition that helps to ensure both sides of the migration can present the same. Therefore it is preferred to specifying the CPU capabilities as a baseline CPU model +/- a list of features to ensure compatibility. In QEMU, it comes down to this example `-cpu` argument:

Reviewer 2: ```suggestion
Libvirt and QEMU allow the CPU to be expressed in ways that ease migration between hosts in heterogeneous infrastructures. While a libvirt VM configured with `host-passthrough` or `maximum` (the equivalents in QEMU are `-cpu host` and `-cpu max` respectively) gets the most out of the hardware, it exposes the guests to all available host capabilities. Therefore, migration to another host is much more restricted, because it needs to be able to present the very same environment to the guest. To do so it needs a clear definition that helps to ensure both sides of the migration can present the same. Therefore it is preferred to specifying the CPU capabilities as a baseline CPU model +/- a list of features to ensure compatibility. In QEMU, it comes down to this example `-cpu` argument:
```

Reviewer 2: "gets the most out of the hardware" - gets the most what?

Reviewer 2: "both sides of the migration can present the same" - the same what? Also, this paragraph feels ike a few points are being repeated a bit. Is there room for us to restructure a bit to avoid repetition?

Reviewer 2: "Therefore it is preferred to specifying the..." -> "It is best to specify..."

```text
-cpu EPYC-Turin,x2apic=on,tsc-deadline=on,hypervisor=on,tsc-adjust=on,
 spec-ctrl=on,stibp=on,flush-l1d=on,ssbd=on,cmp-legacy=on,virt-ssbd=on,
 tsa-sq-no=on,tsa-l1-no=on,pcid=off,la57=off
```

> Or in a level higher, you can ask libvirt to build the CPU for you by using the `host-model` CPU mode.

Reviewer 2: "in a level higher" - in what parameter space are we talking here? Level higher than what?

Author: level higher in the expressiveness of the libvirt VM definition syntax

Reviewer 2: Could we make this a bit clearer? I would never have guessed that that's what that referred to 🙃

> This will detect the current CPU with all its features, but automatically expresses it as baseline CPU model +/- a list of features.

Reviewer 2: ```suggestion
This detects the current CPU with all its features, automatically expressed as baseline CPU model +/- a list of features.
```

```xml
...
  <cpu mode='host-model' check='partial'/>
...
```

> or you might prefer even more control and provide the exact CPU specification:

Reviewer 2: ```suggestion
If you prefer more control, you can provide the exact CPU specification:
```

```xml
...
  <cpu mode='custom' match='exact' check='full'>
    <model fallback='forbid'>EPYC-Turin</model>
    <vendor>AMD</vendor>
    <feature policy='require' name='x2apic'/>
    <feature policy='require' name='tsc-deadline'/>
    ...
    <feature policy='require' name='tsa-l1-no'/>
    <feature policy='disable' name='pcid'/>
    <feature policy='disable' name='la57'/>
...
```

> While the first approach of just using `host-passthrough` and even `host-model` seems easier and more appealing, it is not suitable for heterogeneous infrastructure. To retain full control over migratability across every host in the migration pool a CPU should be modeled.

Reviewer 2: I think it's more helpful for readers if we tell them right from the start which method they should follow so they can skip to the correct section, rather than telling them afterwards that a particular method isn't suitable for certain cases. Otherwise they might waste their time and effort trying something they shouldn't be.

You can check the capabilities of every host in your pool and define the CPU definition to only advertise CPU features to the guest that are guaranteed to exist on every host in the pool.
This creates a stable "lowest common denominator" CPU that looks identical on both the source and the destination, so the guest never comes to rely on a feature that could disappear after a migration.

### Determining the lowest common CPU denominator

Finding the lowest denominator seems to be a complex task. Fortunately, libvirt provides a command to compute it (see {manpage}`virsh(1)` and its `hypervisor-cpu-baseline` command) by taking the CPU definitions of every host in the pool and returning a single baseline CPU that all of them can support.

First, collect the host CPU definition from each host in the pool. The `<cpu>` block reported by `virsh capabilities` describes what the host CPU offers:

> ```text

Reviewer 2: Is this a command? If so, can we render it as a terminal block?

Reviewer 2: (same comment for any other command blocks)

virsh capabilities > host1-caps.xml
```

Then feed all the files to `hypervisor-cpu-baseline`. Libvirt extracts the `<cpu>` element from each file and computes the baseline:

```text
virsh hypervisor-cpu-baseline --migratable host1-caps.xml host2-caps.xml
```

> The `--migratable` option restricts the result to features that are safe to migrate, excluding those that cannot be preserved across a migration. The command prints a `<cpu>` block similar to this:

Reviewer 2: What happens if there are features that can't be preserved? what should the user do if they want those features?

Author: If users want to preserve the CPU features, they would need to restrict the migration to a host that would allow so

```xml
<cpu mode='custom' match='exact'>
  <model fallback='forbid'>EPYC-Turin</model>
  <vendor>AMD</vendor>
  <feature policy='require' name='x2apic'/>
  <feature policy='require' name='tsc-deadline'/>
  ...
</cpu>
```

> This block is the lowest common denominator CPU for the pool. Drop it into the domain XML of the VMs you intend to migrate, so that these VMs see the same CPU regardless of which host they run on.

Reviewer 2: What do we mean by "drop it"? This might be confusing because it conflicts with the commonly used meaning being to discard something

Author: I used "Copy it into" instead of "Drop it"
